sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/nodes/intrinsics/multithreading/UtilFunctionCall.java

```diff
@@ -0,0 +1,178 @@
+        }
+
+        @Override
+        public void run() {
+            synchronized (ctx.callTargetLock) {
+                if (ctx.pthreadCallTarget == null) {
+                    ctx.pthreadCallTarget = Truffle.getRuntime().createCallTarget(new FunctionCallNode(LLVMLanguage.getLanguage()));
+                }
+            }
+            // pthread_exit throws a control flow exception to stop the thread
+            try {
+                // save return value in storage
+                Object returnValue = ctx.pthreadCallTarget.call(startRoutine, arg);
+                // no null values in concurrent hash map allowed
+                if (returnValue == null) {
+                    returnValue = LLVMNativePointer.createNull();
+                }
+                ctx.setThreadReturnValue(Thread.currentThread().getId(), returnValue);
+            } catch (PThreadExitException e) {
+                // return value is written to retval storage in exit function before it throws this
+                // exception
+            } catch (LLVMExitException e) {
+                System.exit(e.getExitStatus());
+            } finally {
+                // call destructors from key create
+                if (this.isThread) {
+                    for (int key = 1; key <= ctx.curKeyVal; key++) {
+                        LLVMPointer destructor = UtilAccessCollectionWithBoundary.get(ctx.destructorStorage, key);
+                        if (destructor != null && !destructor.isNull()) {
+                            ConcurrentMap<Long, LLVMPointer> specValueMap = UtilAccessCollectionWithBoundary.get(ctx.keyStorage, key);
+                            // if key was deleted continue with next destructor
+                            if (specValueMap == null) {
+                                continue;
+                            }
+                            Object keyVal = UtilAccessCollectionWithBoundary.get(specValueMap, Thread.currentThread().getId());
+                            if (keyVal != null) {
+                                // if key value is null pointer continue with next destructor
+                                try {
+                                    LLVMPointer keyValPointer = LLVMPointer.cast(keyVal);
+                                    if (keyValPointer.isNull()) {
+                                        continue;
+                                    }
+                                } catch (Exception e) {
+                                    // ignored
+                                }
+                                UtilAccessCollectionWithBoundary.remove(specValueMap, Thread.currentThread().getId());
+                                new FunctionCallRunnable(destructor, keyVal, this.ctx, false).run();
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    static final class MyArgNode extends LLVMExpressionNode {
```

Comment:
This was just a temporary name I gave some scratch code for demonstration. Please use `NodeFactory#createFreameRead` to create more appropriate nodes instead.

Reply:
ok, but not sure if i did it the right way
i deleted the `MyArgNode` class and used:
`
...
context.getLanguage().getNodeFactory().createFrameRead(PointerType.VOID, this.getFrameDescriptor().findFrameSlot("function"))
...`

i took `PointerType.VOID` for function, arg and stack-pointer.

Reply:
That should work as long as all of these are of pointer type.